src/manifolds/StiefelCanonicalMetric.jl

```diff
@@ -0,0 +1,153 @@
+    A = p' * X
+    n == k && return mul!(q, p, exp(A))
+    QR = qr(X - p * A)
+    BC_ext = exp([A -QR.R'; QR.R 0*I])
+    q .= [p Matrix(QR.Q)] * @view(BC_ext[:, 1:k])
+    return q
+end
+
+@doc raw"""
+    inner(M::MetricManifold{ℝ, Stiefel{n,k,ℝ}, X, CanonicalMetric}, p, X, Y)
+
+compute the inner procuct on the [`Stiefel`](@ref) manifold with respect to the
+[`CanonicalMetric`](@ref). The formula reads
+
+```math
+g_p(X,Y) = \operatorname{tr}\bigl( X^{\mathrm{T}}(I_n - \frac{1}{2}pp^{\mathrm{T}})Y \bigr).
+```
+"""
+function inner(::MetricManifold{ℝ,Stiefel{n,k,ℝ},CanonicalMetric}, p, X, Y) where {n,k}
+    T = Base.promote_eltype(p, X, Y)
+    s = T(dot(X, Y))
+    if n != k
+        s -= dot(p'X, p'Y) / 2
+    end
+    return s
+end
+
+@doc raw"""
+    X = log(M::MetricManifold{ℝ, Stiefel{n,k,ℝ}, CanonicalMetric}, p, q; kwargs..)
+    log!(M::MetricManifold{ℝ, Stiefel{n,k,ℝ}, X, CanonicalMetric}, p, q; kwargs...)
+
+Compute the logarithmic map on the [`Stiefel`](@ref)`(n,k)` manifold with respect to the [`CanonicalMetric`](@ref)
+using a matrix-algebraic based approach to an iterative inversion of the formula of the
+[`exp`](@ref exp(::MetricManifold{ℝ, Stiefel{n,k,ℝ}, CanonicalMetric}, ::Any...) where {n,k}).
+
+The algorithm is derived in[^Zimmermann2017].
+
+# Keyword arguments
+
+Both `maxiter`(`=1e5`) and `tolerance`(`=1e-9`) can be given to the iterative process as
+stoppping criteria.
+
+[^Zimmermann2017]:
+    > Zimmermann, R.: _A matrix-algebraic algorithm for the Riemannian logarithm on the Stiefel manifold under the canoncial metric.
+    > SIAM Journal on Matrix Analysis and Applications 28(2), pp. 322-342, 2017.
+    > doi: [10.1137/16M1074485](https://doi.org/10.1137/16M1074485),
+    > arXiv: [1604.05054](https://arxiv.org/abs/1604.05054).
+"""
+log(::MetricManifold{ℝ,Stiefel{n,k,ℝ},CanonicalMetric}, ::Any...) where {n,k}
+
+function log(
+    M::MetricManifold{ℝ,Stiefel{n,k,ℝ},CanonicalMetric},
+    p,
+    q;
+    maxiter=1e5,
+    tolerance=1e-9,
+) where {n,k}
+    X = allocate_result(M, log, p, q)
+    log!(M, X, p, q; maxiter=maxiter, tolerance=tolerance)
+    return X
+end
+
+function log!(
+    ::MetricManifold{ℝ,Stiefel{n,k,ℝ},CanonicalMetric},
+    X,
+    p,
+    q;
+    maxiter=1e5,
+    tolerance=1e-9,
+) where {n,k}
+    M = p' * q
+    QR = qr(q - p * M)
+    V = Matrix(qr([M; QR.R]).Q * I)
+    Vcorner = @view V[(k + 1):(2 * k), (k + 1):(2k)] #bottom right corner
+    Vpcols = @view V[:, (k + 1):(2 * k)] #second half of the columns
+    S = svd(Vcorner) # preprocessing: Procrustes
+    mul!(Vpcols, Vpcols * S.U, S.V')
+    V[:, 1:k] .= [M; QR.R]
+
+    LV = real.(log(V)) # this can be replaced by log_safe.
+    C = view(LV, (k + 1):(2 * k), (k + 1):(2 * k))
+    expnC = exp(-C)
+    i = 0
+    while (i < maxiter) && (norm(C) > tolerance)
+        i = i + 1
+        LV .= real.(log(V))
+        expnC .= exp(-C)
+        copyto!(Vpcols, Vpcols * expnC)
```

Comment:
Can you allocate a temporary outside the loop and then multiply into it? e.g. `mul!(tmp, Vpcols, expnC)`